minio/progress.py

```diff
@@ -0,0 +1,96 @@
+import time
+import sys
+from threading import Thread
+from .compat import queue, queue_empty
+
+
+class Progress(Thread):
+    def __init__(self, total_size, file_name, interval=1):
+        Thread.__init__(self)
+        self.daemon = True
+        self.total_size = total_size
+        self.display_queue = queue()
+        self.current_size = 0
+        self.file_name = file_name
+        self.prefix = self.file_name + ': ' if self.file_name else ''
+        self.sp = StatusPrinter(sys.stdout)
+        self.start_t = time.time()
+        self.interval = interval
+
+    def run(self):
+        displayed_time = 0
+
+        while True:
+            try:
+                # display every 1 secs
+                task = self.display_queue.get(timeout=1)
+            except queue_empty:
+                elapsed_time = time.time() - self.start_t
+                if elapsed_time > displayed_time:
+                    displayed_time = elapsed_time
+                self.sp.print_status(
+                    self.prefix + format_meter(self.current_size, self.total_size, displayed_time))
+                continue
+            prefix, now, total = task
+            displayed_time = time.time() - self.start_t
+            self.sp.print_status(self.prefix + format_meter(now, total, displayed_time))
+            self.display_queue.task_done()
+            if now == total:
+                break
+
+    def update(self, size):
+
+        if self.current_size == 0:
+            self.start()
+            self.display_queue.put((self.prefix, 0, self.total_size))
+
+        self.current_size += size
+        self.display_queue.put((self.prefix, self.current_size, self.total_size))
+        if self.current_size == self.total_size:
+            self.display_queue.join()
+
+
+def format_interval(t):
+    mins, s = divmod(int(t), 60)
+    h, m = divmod(mins, 60)
+    if h:
+        return '%d:%02d:%02d' % (h, m, s)
+    else:
+        return '%02d:%02d' % (m, s)
+
+
+def format_meter(n, total, elapsed):
+    # n - number of finished iterations
+    # total - total number of iterations, or None
+    # elapsed - number of seconds passed since start
+    if n > total:
+        total = None
+
+    elapsed_str = format_interval(elapsed)
+    n_to_mb = n / 1024 / 1024
```

Comment:
Define all the constants, constants should be defined in the beginning of this file